docs: reasoning and use cases for channel upgradability (#5346)
cosmos/ibc-go: docs/docs/01-ibc/06-channel-upgrades.md

Learn how to upgrade existing IBC channels.
:::

Channel upgradeability is an IBC-level protocol that allows chains to leverage new channel features without having to create new channels or perform a network-wide upgrade. 

Prior to this feature, developers who wanted to update an application module or add a middleware to their application flow would need to negotiate a new channel in order to use the updated application feature/middleware, resulting in a loss of the accumulated state/liquidity, token fungibility (as the channel would have been encoded in the IBC denom), and any other larger network effects of losing usage of the existing channel from relayers monitoring, etc.

> With channel upgradeability, applications will be able to implement features such as but not limited to: [including a memo field in the packet data for fungible tokens](https://github.com/cosmos/ibc/pull/842), adding [denom metadata to tokens](https://github.com/cosmos/ibc/discussions/719), or utilizing the [fee middleware](https://github.com/cosmos/ibc/tree/main/spec/app/ics-029-fee-payment), all while maintaining the channels on which they currently operate.

Reviewer: `including a memo field in the packet data for fungible tokens](https://github.com/cosmos/ibc/pull/842)` this should be possible without channel upgradability, because the memo was a backwards-compatible addition, right?

`[denom metadata to tokens](https://github.com/cosmos/ibc/discussions/719)` should we add a disclaimer that is not implemented yet though.

This document outlines the channel upgrade feature, and the multiple steps used in the upgrade process.

## Channel Upgrade Handshake

Channel upgrades will be initialized using a handshake process that is designed to be similar to the standard connection/channel opening handshake.

```go
interface ChannelEnd {
  state: ChannelState
  ordering: ChannelOrder
  counterpartyPortIdentifier: Identifier
  counterpartyChannelIdentifier: Identifier
  connectionHops: [Identifier]
  version: string
  upgradeSequence: uint64

> }

Reviewer: Instead of the interface from the spec, should we use here the proto or go definition?

```

Anything within this channel end interface except the identifiers (counterparty port ID and channel ID) can be changed — such as the version, connection hops, and channel ordering. For example, the fee middleware can be added to an application module by updating the version string [shown here](https://github.com/cosmos/ibc-go/blob/995b647381b909e9d6065d6c21004f18fab37f55/modules/apps/29-fee/types/metadata.pb.go#L28).

There a few caveats for the changes in the channel end fields:

1. Connection hops can change in a channel upgrade, however both sides must still be each other's counterparty.
2. The channel ordering can only be upgraded if the new ordering is a subset of its previous one. This means that a channel can be upgraded from an ordered to an unordered channel but not the other way around.

On a high level, successful handshake process for channel upgrades works as follows:

1. The chain initiating the upgrade process (chain A) sets its channel state from `OPEN` to `INITUPGRADE` via the `ChanUpgradeInit` function.

> 2. Subsequently, the counterparty (chain B) changes its channel end from `OPEN` to `TRYUPGRADE` with `ChanUpgradeTry`.

Author: argh, this is what i get from copying from the spec 🙈

Learn how to upgrade existing IBC channels.
:::

> Channel upgradability is an IBC-level protocol that allows chains to leverage new channel features without having to create new channels or perform a network-wide upgrade. 

Reviewer 2: ```suggestion
Channel upgradability is an IBC-level protocol that allows chains to leverage new application and channel features without having to create new channels or perform a network-wide upgrade. 
```

> Prior to this feature, developers who wanted to update an application module or add a middleware to their application flow would need to negotiate a new channel in order to use the updated application feature/middleware, resulting in a loss of the accumulated state/liquidity, token fungibility (as the channel would have been encoded in the IBC denom), and any other larger network effects of losing usage of the existing channel from relayers monitoring, etc.

Reviewer 3: ```suggestion
Prior to this feature, developers who wanted to update an application module or add a middleware to their application flow would need to create a new channel in order to use the updated application feature/middleware, resulting in a loss of the accumulated state/liquidity, token fungibility (as the channel ID is encoded in the IBC denom), and any other larger network effects of losing usage of the existing channel from relayers monitoring, etc.
```

With channel upgradability, applications will be able to implement features such as but not limited to: potentially adding [denom metadata to tokens](https://github.com/cosmos/ibc/discussions/719), or utilizing the [fee middleware](https://github.com/cosmos/ibc/tree/main/spec/app/ics-029-fee-payment), all while maintaining the channels on which they currently operate.

This document outlines the channel upgrade feature, and the multiple steps used in the upgrade process.

## Channel Upgrade Handshake

> Channel upgrades will be initialized using a handshake process that is designed to be similar to the standard connection/channel opening handshake.

Reviewer 2: mega nit tbf

```suggestion
Channel upgrades will be achieved using a handshake process that is designed to be similar to the standard connection/channel opening handshake.
```

```go
type Channel struct {
  // current state of the channel end
  State State `protobuf:"varint,1,opt,name=state,proto3,enum=ibc.core.channel.v1.State" json:"state,omitempty"`
  // whether the channel is ordered or unordered
  Ordering Order `protobuf:"varint,2,opt,name=ordering,proto3,enum=ibc.core.channel.v1.Order" json:"ordering,omitempty"`
  // counterparty channel end
  Counterparty Counterparty `protobuf:"bytes,3,opt,name=counterparty,proto3" json:"counterparty"`
  // list of connection identifiers, in order, along which packets sent on
  // this channel will travel
  ConnectionHops []string `protobuf:"bytes,4,rep,name=connection_hops,json=connectionHops,proto3" json:"connection_hops,omitempty"`
  // opaque channel version, which is agreed upon during the handshake
  Version string `protobuf:"bytes,5,opt,name=version,proto3" json:"version,omitempty"`
  // upgrade sequence indicates the latest upgrade attempt performed by this channel
  // the value of 0 indicates the channel has never been upgraded
  UpgradeSequence uint64 `protobuf:"varint,6,opt,name=upgrade_sequence,json=upgradeSequence,proto3" json:"upgrade_sequence,omitempty"`
}
```

> The version, connection hops, and channel ordering are fields in this channel struct which can be changed. For example, the fee middleware can be added to an application module by updating the version string [shown here](https://github.com/cosmos/ibc-go/blob/995b647381b909e9d6065d6c21004f18fab37f55/modules/apps/29-fee/types/metadata.pb.go#L28). However, although connection hops can change in a channel upgrade, both sides must still be each other's counterparty.

Reviewer 2: ```suggestion
The version, connection hops, and channel ordering are fields in this channel struct which can be changed. For example, the fee middleware can be added to an application module by updating the version string [shown here](https://github.com/cosmos/ibc-go/blob/995b647381b909e9d6065d6c21004f18fab37f55/modules/apps/29-fee/types/metadata.pb.go#L28). However, although connection hops can change in a channel upgrade, both sides must still be each other's counterparty. This is enforced by the upgrade protocol and upgrade attempts which try to alter an expected counterparty will fail.
```

On a high level, successful handshake process for channel upgrades works as follows:

> 1. The chain initiating the upgrade process will propose a potential upgrade.

Reviewer 2: nit, could drop "potential"

```suggestion
1. The chain initiating the upgrade process will propose an upgrade.
```

> 1. The chain initiating the upgrade process (chain A) submits the `ChanUpgradeInit` function.

Reviewer 3: ```suggestion
1. The chain initiating the upgrade process (chain A) executes the `ChanUpgradeInit` handshake step.
```

> 2. If the counterparty agrees with the proposal, it will block sends and begin flushing its channel end. 

Reviewer 2: maybe we should be a bit more explicit here about "flushing" since it looks like this is the first time the language is introduced

```suggestion
2. If the counterparty agrees with the proposal, it will block sends and begin flushing any in-flight packets on its channel end. 
```

Author: i see what you mean, will add that flushing will be detailed later specifically!

3. Upon successful completion of the previous step, the initiating chain will also block packet sends and begin flushing its channel end. 

> 4. Once both channel ends have completely flushed before either upgrade timeout has elapsed, both channel ends can be opened and upgraded to the new channel fields. 

Reviewer 2: ```suggestion
4. Once both channel ends have completed flushing packets within the upgrade timeout window, both channel ends can be opened and upgraded to the new channel fields. 
```

> 1. The chain initiating the upgrade process (chain A) submits the `ChanUpgradeInit` function.
> 2. Subsequently, the counterparty (chain B) changes its channel end from `OPEN` to `FLUSHING` with `ChanUpgradeTry`.
> 3. Upon successful completion of the previous step, chain A sets its channel state from `OPEN` to `FLUSHING` or `FLUSHCOMPLETE` with `ChanUpgradeAck` depending on the packet flushing status
> 4. Finally, chain B sets its channel state to `FLUSHCOMPLETE` when its packet flush is finished, and then `OPEN` with `ChanUpgradeConfirm`.

Reviewer 3: I wonder if we should make it even more high level:
```suggestion
1. The chain initiating the upgrade process will propose a potential upgrade.
2. If the counterparty agrees with the proposal, it will block sends and begin flushing its channel end. 
3. Upon successful completion of the previous step, the initiating chain will also block packet sends and begin flushing its channel end. 
4. Once both channel ends have completely flushed before either upgrade timeout has elapsed, both channel ends can be opened and upgraded to the new channel fields. 
```

Reviewer 3: I like this a bit better, because I think it's hard to capture each of the 5 handshake steps in a brief summary
